addOns/graphql/src/main/java/org/zaproxy/addon/graphql/GraphQlFingerprinter.java

```diff
@@ -0,0 +1,570 @@
+    }
+
+    private boolean checkGraphQlByPopEngine() {
+        sendQuery("{alias1$1:__typename}");
+        if (lastQueryMsg != null && lastQueryMsg.getResponseHeader().isJson()) {
+            try {
+                String response = lastQueryMsg.getResponseBody().toString();
+                JsonElement data = GSON.fromJson(response, JsonObject.class).get("data");
+                if (data != null && data.isJsonObject()) {
+                    var dataObject = data.getAsJsonObject();
+                    if (dataObject.has("alias1$1")
+                            && "QueryRoot".equals(dataObject.get("alias1$1").getAsString())) {
+                        matchedString = "QueryRoot";
+                        return true;
+                    }
+                }
+            } catch (Exception ignored) {
+            }
+        }
+        sendQuery("query aa#aa {__typename}");
+        if (errorContains("Unexpected token \"END\"")) {
+            return true;
+        }
+        sendQuery("query @skip {__typename}");
+        if (errorContains("Argument 'if' cannot be empty, so directive 'skip' has been ignored")) {
+            return true;
+        }
+        sendQuery("query @doesnotexist {__typename}");
+        if (errorContains("No DirectiveResolver resolves directive with name 'doesnotexist'")) {
+            return true;
+        }
+        sendQuery("");
+        return errorContains("The query in the body is empty");
+    }
+
+    private boolean checkGraphQlGoEngine() {
+        sendQuery("{__typename{}");
+        if (errorContains("Unexpected empty IN")) {
+            return true;
+        }
+        sendQuery("");
+        if (errorContains("Must provide an operation.")) {
+            return true;
+        }
+        sendQuery("{__typename}");
+        try {
+            String response = lastQueryMsg.getResponseBody().toString();
+            JsonElement data = GSON.fromJson(response, JsonObject.class).get("data");
+            if (data != null && data.isJsonObject()) {
+                var dataObject = data.getAsJsonObject();
+                if (dataObject.has("__typename")
+                        && "RootQuery".equals(dataObject.get("__typename").getAsString())) {
+                    matchedString = "RootQuery";
+                    return true;
+                }
+            }
+        } catch (Exception ignored) {
+        }
+        return false;
+    }
+
+    private boolean checkGraphQlJavaEngine() {
+        sendQuery("queryy {__typename}");
+        if (errorContains("Invalid Syntax : offending token 'queryy'")) {
+            return true;
+        }
+        sendQuery("query @aaa@aaa {__typename}");
+        if (errorContains(
+                "Validation error of type DuplicateDirectiveName: Directives must be uniquely named within a location.")) {
+            return true;
+        }
+        sendQuery("");
+        return errorContains("Invalid Syntax : offending token '<EOF>'");
+    }
```

Comment:
It seems like 80% of the methods could be generalized to accept a map of query/error and run through a generic looping method (works for one works for many)? 🤷‍♂️

Reply:
I'd say it's closer to 60% because many methods include additional checks. I'm happy to refactor the code but how would that help? Is it better for readability, or more efficient, or something else?

Reply:
I just figured it would reduce duplication (the amount of code to be maintained).

Reply:
That's fair, but while that may be less code, I feel it will be less uniform than the current code, since there would still be methods which need to follow the current pattern. Also, we would be initializing a Map for checks that call the looping method and that may be slightly more expensive?

Reply:
Okay, I don't want to block on this.

I just figured that instead of passing a custom method for each framework we could generalize and parameterized one for a bunch, which may make future additions easier too :shrug: